storage/remote/queue_manager.go

```diff
@@ -946,49 +1069,69 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue chan sample) {
 			// Number of pending samples is limited by the fact that sendSamples (via sendSamplesWithBackoff)
 			// retries endlessly, so once we reach max samples, if we can never send to the endpoint we'll
 			// stop reading from the queue. This makes it safe to reference pendingSamples by index.
-			pendingSamples[nPending].Labels = labelsToLabelsProto(sample.labels, pendingSamples[nPending].Labels)
-			pendingSamples[nPending].Samples[0].Timestamp = sample.t
-			pendingSamples[nPending].Samples[0].Value = sample.v
-			nPending++
+			switch d := sample.(type) {
+			case writeSample:
+				sampleBuffer[nPendingSamples][0] = d.sample
+				pendingSamples[nPending].Labels = labelsToLabelsProto(d.seriesLabels, pendingSamples[nPending].Labels)
+				pendingSamples[nPending].Samples = sampleBuffer[nPendingSamples]
+				pendingSamples[nPending].Exemplars = nil
+				nPendingSamples++
+				nPending++
+
+			case writeExemplar:
+				exemplarBuffer[nPendingExemplars][0] = d.exemplar
+				pendingSamples[nPending].Labels = labelsToLabelsProto(d.seriesLabels, pendingSamples[nPending].Labels)
```

Comment:
I think it's fine for first iteration, but I would add GH issue to describe this simplicity. I am only worried this is not the most efficient way, and server implementations will rely on this fact and it will be hard to change 🤔

Reply:
Correct. We don't bundle samples and exemplars for the same series in a single `TimeSeries`.

`Also this means that when tailing we don't batch samples together, ever too?`
Also correct. We could try and do something here, but the main concern would be limiting memory usage.

Reply:
👍🏽